keps/0002-kong-gateway-api.md

```diff
@@ -13,23 +13,271 @@ status: provisional
+    gateways, err := s.ListGateways()
+    parsedGateways := fromGatewayAPIsV1alpha2Gateway(log, gateways)
+    httproutes, err := s.ListHTTPRoutes()
+    parsedHTTPRoutes := fromGatewayAPIsV1alpha2HTTPRoute(log, httproutes)
+
+    return mergeIngressRules(otherTypes..., parsedGateways, parsedHTTPRoutes)
+```
+
+The parsing logic above (e.g. `fromGatewayAPIsV1alpha2HTTPRoute()`) is the ultimate low-level logic we will need to compile `Gateways` and `HTTPRoutes`
+into `kong.Services` and `kong.Routes`. Practically speaking almost all the prior art needed to implement this is covered by the API types we already
+support, so the work should be predominantly integration work and little to no exploratory work needed.
+
+### Test Plan
+
+In order to consider this KEP `implemented` we must have the following present:
+
+- `examples/` includes manifests for (at least) a working HTTP ingress with `GatewayClass`, `Gateway`, and `HTTPRoute` using Kong
+- KIC [docs.konghq.com][docs] must include a `guide/` for Gateway APIs (flagged **experimental** until Gateway APIs are GA)
+- `test/integration/` includes integration tests which cover deployment and configuration options of `HTTPRoute`, including TLS configurations
+- our `.github/workflows/release.yaml` release CI must test our `HTTPRoute` implementation against a historical list of the Kubernetes versions we support
+
+[docs]:https://docs.konghq.com/kubernetes-ingress-controller/
+
+## Production Readiness
+
+Production readiness of this feature is marked by the following requirements:
+
+- Gateway APIs themselves have reached a GA status in upstream Kubernetes
+- A version of Kubernetes which includes the Gateway APIs standard becomes GA
+- We have support for most if not all of the [entire Gateway APIs spec][gateway-spec]
+- If there's any features or implementations for Gateway APIs we explicitly choose not to support, we document that and the reasoning
+- Our integration and E2E testing provides feature cover and strong regression protections
+
+[gateway-spec]:https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/
+
+### Feature Enablement and Rollback
+
+In order to significantly limit the exposure of our Gateway APIs implementations during the alpha/beta periods of their release
+we will not provide the standard `--enable-controller-<name>` flags for the manager as these are [prominently documented][flags-docs]
+and otherwise easily visible when running `manager --help` which ostensibly gives them a sense that they are a normal part of any included release.
+
+We will provide a [Feature Gate][feature-gates] option for enabling Gateway APIs which will [conform to upstream standards][gate-alpha-beta] as to how
+alpha and beta features can be enabled, which in turn will be strongly familiar way to manage these kinds of features to Kubernetes cluster operators.
+
+Consequently we will need to add [documentation][docs-refs] a table of our feature gates statuses as upstream does for their controllers.
```

Comment:
Maybe unnecessary--I think that if these are on, it makes sense that you'd have the same flags as the other controllers. If you can even create Gateway resources, it implies you've enabled the feature flag in other, core, components (e.g. kube-apiserver) already.

We can leverage that and use logic similar to our other maybe controllers: if the Gateway resources are available on the API server, and you have not explicitly disabled their controller, we load that controller.

Reply:
The controller flags and mechanisms we have now for enabling API support do not make distinctions about the maturity of the featuresets. I specifically want to be able to release gateway support with very specific (and upstream familiar) documented maturity, as well as make it very hard for someone who has not read the documentation and has been strongly warned about the maturity of that feature to find out how to turn that feature on without having to read those warnings.

Reply:
I'm going to remove the feature gating components from this KEP for now because it doesn't seem like we're on the same page on that, but we are on other things. I'll try to approach feature gates as its own thing (perhaps its own KEP) separately.

Reply:
bebe6a59fbe384a0562df390e0223b38dd7b4fcf